library/core/src/slice/mod.rs

```diff
@@ -90,11 +90,23 @@ impl<T> [T] {
     #[rustc_const_stable(feature = "const_slice_len", since = "1.32.0")]
     #[inline]
     // SAFETY: const sound because we transmute out the length field as a usize (which it must be)
-    #[allow_internal_unstable(const_fn_union)]
+    #[allow_internal_unstable(const_fn_union, const_unreachable_unchecked)]
     pub const fn len(&self) -> usize {
         // SAFETY: this is safe because `&[T]` and `FatPtr<T>` have the same layout.
         // Only `std` can make this guarantee.
-        unsafe { crate::ptr::Repr { rust: self }.raw.len }
+        let raw_len = unsafe { crate::ptr::Repr { rust: self }.raw.len };
+
+        match mem::size_of::<T>() {
+            0 => {}
+            // SAFETY: this branch taken if mathematically `isize::MAX < raw_len * N`.
+            // But references must point to one allocation with size at most isize::MAX.
+            n if (isize::MAX as usize) / n < raw_len => unsafe {
```

Comment:
Why do you prefer the division over multiplication like the comment above explained ?

Reply:
The division is a compile time constant so the branch is exactly the restrictino of the dynamically loaded value to a bound. I don't think the multiplication is equivalent. From the viewpoint of a compiler that does _not_ already assume the range semantics, the multiplication could still overflow and be accidentally correct. It is only mathematically the intended check.

Reply:
I believe compiler/LLVM could see what constants and perform needed optimization.
But yeah, don't mind me 'cause I might be wrong.

Reply:
Here's a playground demonstrating it wouldn't work on `size_of::<T>() > 1`:

https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=5e3cc651f2d4121ea74b04251d211a2f

The division however works. This should be added to the tests, and to the comments.